### DISCUSSION TITLE
Oauth2 Multiple URLs (exclude localhost)

### DISCUSSION THREAD

**Discussion by Firstyear** [kanidm/kanidm#2511] [Ideas]
We have had a number of past issues asking for multiple url support in oauth2. So far almost all of these requests have been about allowing an RP to have a https://example.com name while also allowing redirects to http://localhost. 

However, there may be a case where someone wants to allow multiple names such as https://au.example.com and https://nz.example.com.

We need to understand *why* someone may want multiple URLs like this. From our app portal page, we can only nominate a single outbound link, which may constrain possible deployment options when multiple urls are involved.

This issue is intended to serve as a discussion about multiple public urls only. We need to know how or why someone may deploy these so that we can properly implement a design to facilitate this.

**Comment by 0xC0ncord** [comment]
One use case for this is to have multiple environments where the application is deployed but only one single OAuth2 client, e.g. a production, staging, and dev environment:
- myapp.example.com
- myapp-staging.example.com
- myapp-dev.example.com

**Comment by Firstyear** [comment]
> One use case for this is to have multiple environments where the application is deployed but only one single OAuth2 client, e.g. a production, staging, and dev environment:
> 
>     * myapp.example.com
> 
>     * myapp-staging.example.com
> 
>     * myapp-dev.example.com

This use-case is *exactly* why I don't want to allow multiple origins. You should never mix production staging and dev credentials and application domains. This would allow a dev access token to be usable against a production resource server. So compromise of dev becomes compromise of production.

At the moment by forcing single origin, we force you to have separate security domains between each client, preventing users from making the security mistake in the first place.

**Reply by someone-stole-my-name** [reply]
Agree on never mixing credentials between stages. However, this is still a valid use-case if you have N _distinct_ but equal environments, e.g. multiple and completely isolated staging environments:

* myapp-dev-1.example.com
* myapp-dev-2.example.com

Our use case is on-demand staging environments, where all of them are equal for the purposes of kanidm but under different domains.

**Comment by 0xC0ncord** [comment]
>This use-case is exactly why I don't want to allow multiple origins. You should never mix production staging and dev credentials and application domains. This would allow a dev access token to be usable against a production resource server. So compromise of dev becomes compromise of production.

>At the moment by forcing single origin, we force you to have separate security domains between each client, preventing users from making the security mistake in the first place.

I agree I think that's a very compelling reason to not have such a capability. Truthfully I can't think of another use case myself -- I don't personally have a need for this feature. I was searching around for potential use cases based on the fact that Keycloak supports this but the reasoning I provided was the only one I could find.

**Comment by Firstyear** [comment]
Yeah. The previous usecase that almost all requests were related too was to have public clients with localhost redirects - we've solved that now in a more secure, and simpler manner.

**Comment by TheRealGramdalf** [comment]
I'm really not sure how this is intended to work, but would multiple URLs/URIs be required for e.g. https://github.com/astubenbord/paperless-mobile/issues/374 (using a `paperless://` URI with a mobile application)? Or does that only concern the callback URI, since actual authentication happens directly in a web page (which uses the correct URL)?

**Comment by yaleman** [comment]
> I'm really not sure how this is intended to work, but would multiple URLs/URIs be required for e.g. [astubenbord/paperless-mobile#374](https://github.com/astubenbord/paperless-mobile/issues/374) (using a `paperless://` URI with a mobile application)? Or does that only concern the callback URI, since actual authentication happens directly in a web page (which uses the correct URL)?

That *could* be a valid use case, but we'd need to see how it worked in general (ie, can it be tested with an RS dedicated to that URL?)

.. though most likely the auth will be against the server's URL, which the app can request a token for using that URL.

**Comment by pando85** [comment]
For example I'm setting up Immich and it requires three URIs:
- app.immich:/ - for logging in with OAuth from the [Mobile App](https://immich.app/docs/features/mobile-app)
- http://DOMAIN:PORT/auth/login - for logging in with OAuth from the Web Client
- http://DOMAIN:PORT/user-settings - for manually linking OAuth in the Web Client

Source: [https://immich.app/docs/administration/oauth#prerequisites]()

**Reply by yaleman** [reply]
Given the [mobile-login-redirect is built in](https://immich.app/docs/administration/oauth/#mobile-redirect-uri), I'm not sure why they even require the app schema login as a redirect point TBH - I wouldn't be supporting that first entry as a valid option.

**Reply by pando85** [reply]
Yes, I'm not sure either. But there are an alternative to use a different URI. How could I achieve that in Kanidm?
I only know how to setup a redirect URL and mobile and web need two different endpoints.

**Reply by yaleman** [reply]
The point of this discussion is to find any use case that actually used multiple redirect URLs before implementing the feature, this one's finally one of them by the looks of it.

**Reply by Firstyear** [reply]
Currently do we actually support non-http urls anyway? 

I think we could even end up supporting this via the future "multi-client" oauth2 feature I have in mind.

**Reply by pando85** [reply]
That's not even critical because you can change the URL endpoint for the mobile app and it can be http.

Could I implement the multiple redirect URLs? Do you have something in mind for this feature which is supported in different IDMs with Oauth2/OIDC support?

Also, I found other applications that need this like ArgoCD CLI which is a crucially app for any organization based on GitOps.

**Reply by Firstyear** [reply]
Currently we've been working to improve how we handle keys server side, which will allow us to associate multiple clients to a "oauth2 client domain" where they all share the same keys. This way, we can have two clients (one for each url) where the tokens are valid for both clients and resource servers. That way we retain the current and same logic, and you get a clearer distinction between which url is associate to which client, and we retain our strict logic about redirect flows

**Reply by pando85** [reply]
Ok, that will work.

Is that currently implemented in the last dev version?
Just to be clear, using just one client ID (in the application) we could use any endpoint associated to the shared clients, no? 

It looks like a good solution for me.

**Reply by Firstyear** [reply]
Not yet, it's on the todo list :)

**Reply by pando85** [reply]
Do you have the steps that are needed to get this thing done? I could try to push for it because I'm interested in this feature.

I would like to know mainly what do you expect from the implementation if is possible.

**Reply by Firstyear** [reply]
https://github.com/kanidm/kanidm/pull/2867

**Comment by Firstyear** [comment]
Will be in next release.

**Comment by farcaller** [comment]
I think I have a valid use case that is even somewhat covered in the [kanidm docs](https://kanidm.github.io/kanidm/stable/examples/kubernetes_ingress.html)—cooperating with oauth2-proxy.

I have several http ingresses all behind the same oauth2-proxy instance that is singlehandedly talking to kanidm. Based on which domain you will hit first without the auth, the origin changes. Ideally, oauth2-proxy would do a redirect jump from `myapp.example.com` to `auth.example.com` to then `kanidm.example.com`, but that's just not how it works, it will go straight from `myapp.example.com` or `myapp2.example.com` to kanidm.

With Auth0, I can use `https://*.example.com` as my origin. I will end up with the same JWT for myapp and myapp2, but in my case it's fine as the access scope is similar.

**Reply by Firstyear** [reply]
myapp and myapp2 two are separate apps, and so require separate configurations. I don't know how much more i need to say it, but you can not and should not be sharing "client credentials" for oauth2 with a catch all like *.example.com because you make a huge failure domain with massive blast radius.

**Reply by farcaller** [reply]
The point is that credentials are *not* shared. The only app that can access the credentials is oauth2-proxy, it just so happens to be served from several distinct domains.

I wouldn’t be concerned about the blast radius in this case, because the authentication here is done only against the single app, and that app performs authorization (is the JWT ok enough to access this group of internal apps). The actual protected apps never have access to the token, as it's either stored in oauth2-proxy's local session or in an encrypted cookie.

I don’t think this breaks security preconditions. That said, the PR mentioned above is reasonably sufficient for this case anyway.

**Reply by Firstyear** [reply]
You misunderstand. The issue is that *tokens* issued to a client can be used against all the *resource servers*. So compromise of one resource server compromises *every other resource server*. So if you have app A and app B behind some proxy, some exploit of app A allows re-use of tokens to B. Similar if the tokens are forwarded anywhere. So we strongly discourage this setup because of the risk of lateral movement.

**Reply by farcaller** [reply]
I get your point, but again, I don’t think it's applicable in this case. The resource applications A and B never get the tokens. The only party dealing with the oauth flow is oauth2-proxy (and your ingress doing the routing, of course). This way, you can only compromise oauth2-proxy itself to access the applications guarded by it, but it's not a lateral movement per se, it's more like a single application deployed to several domains.

**Comment by Danny-Dunn** [comment]
Hi, I have an additional use case: Proxmox. In a Proxmox cluster, all of the hosts must share an OAuth client configuration. All hosts in the cluster must function as the same app. In this case you are authenticating to the same environment, just accessed from different hostnames.

**Reply by yaleman** [reply]
This is now doable with add-redirect-url.

**Reply by Danny-Dunn** [reply]
> This is now doable with add-redirect-url. 

My bad, I misunderstood the usage of that option.
